src/Microsoft.DocAsCode.Build.Engine/Incrementals/ChangeList.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.Engine.Incrementals
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.Exceptions;
+    using Microsoft.DocAsCode.Utility;
+
+    public sealed class ChangeList
+    {
+        private readonly List<ChangeItem> _list = new List<ChangeItem>();
+
+        public static ChangeList Parse(string tsvFile, string baseDir)
+        {
+            if (tsvFile == null)
+            {
+                throw new ArgumentNullException(nameof(tsvFile));
+            }
+            if (baseDir == null)
+            {
+                throw new ArgumentNullException(nameof(baseDir));
+            }
+            if (File.Exists(tsvFile))
+            {
+                throw new FileNotFoundException("File not found.", tsvFile);
+            }
+            return ParseCore(tsvFile, baseDir);
+        }
+
+        public void Add(string filePath, ChangeKind kind)
+        {
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+            if (filePath.Length == 0)
+            {
+                throw new ArgumentException("File path cannot be empty", nameof(filePath));
+            }
+            if (!PathUtility.IsRelativePath(filePath))
+            {
+                throw new ArgumentException("Expect relative path.", nameof(filePath));
+            }
+            AddCore(filePath, kind);
+        }
+
+        public IEnumerable<string> GetCreatedFiles() =>
+            from item in _list
+            where item.Kind == ChangeKind.Created
+            select item.FilePath;
+
+        public IEnumerable<string> GetUpdatedFiles() =>
+            from item in _list
+            where item.Kind == ChangeKind.Updated
+            select item.FilePath;
+
+        public IEnumerable<string> GetDeletedFiles() =>
+            from item in _list
+            where item.Kind == ChangeKind.Deleted
+            select item.FilePath;
+
+        private void AddCore(string filePath, ChangeKind kind)
+        {
+            _list.Add(new ChangeItem { FilePath = filePath, Kind = kind });
+        }
+
+        private static ChangeList ParseCore(string tsvFile, string baseDir)
+        {
+            var result = new ChangeList();
+            bool hasError = false;
+
+            foreach (var line in File.ReadLines(tsvFile))
```

Comment:
never be null, see: 
[http://referencesource.microsoft.com/#mscorlib/system/io/file.cs](http://referencesource.microsoft.com/#mscorlib/system/io/file.cs)   line 1036
and
[http://referencesource.microsoft.com/#mscorlib/system/io/ReadLinesIterator.cs](http://referencesource.microsoft.com/#mscorlib/system/io/ReadLinesIterator.cs)  line 50

---
In reply to: [73996456](https://github.com/dotnet/docfx/pull/586#discussion_r73996456) [](ancestors = 73996456)